rclcpp/include/rclcpp/subscription_factory.hpp

```diff
@@ -117,6 +119,32 @@ create_subscription_factory(
       return sub_base_ptr;
     };
 
+  factory.create_typed_subscription_intra_process =
+    [allocator, any_subscription_callback](
+    rclcpp::IntraProcessBufferType buffer_type,
+    rclcpp::Context::SharedPtr context,
+    const std::string & topic_name,
+    const rcl_subscription_options_t & subscription_options
+    ) -> rclcpp::SubscriptionIntraProcessBase::SharedPtr
+    {
+      // If the user has not specified a type for the intra-process buffer, use the callback one.
+      if (buffer_type == IntraProcessBufferType::CallbackDefault) {
+        buffer_type = any_subscription_callback.use_take_shared_method() ?
+          IntraProcessBufferType::SharedPtr : IntraProcessBufferType::UniquePtr;
+      }
+
+      auto sub_intra_process =
+        std::make_shared<SubscriptionIntraProcess<MessageT, Alloc, std::default_delete<MessageT>, CallbackMessageT>>(
```

Comment:
Classes taking an allocator type as template parameter don't need a deleter too (the deleter can be deduced from the allocator type, e.g.: https://github.com/ros2/rclcpp/blob/9be3e08cd4d04193fb76cb6dd8f5db11d7199e2f/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L64-L66
A deleter template parameter is needed in functions with an `unique_ptr` argument (or return value) that takes it (it could be sth different of a `unique_ptr` too, any class needing a deleter), or in a class that needs to know about the deleter and not the allocator type (which I think doesn't happen here).
You usually don't need to pass the deleter template parameter explicitly (because it's automatically deduced). This case is probably an exception of that, so pass a deleter deduced from the allocator type (`create_subscription_factory` is a template taking an allocator type) and not `std::default_delete`.

Reply:
Updated the template argument.. I'm not very familiar with C++ allocator concepts, sorry that it required so many changes.

Reply:
Allocator usage looks correct now.